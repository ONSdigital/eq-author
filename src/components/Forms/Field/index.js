import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

const StyledField = styled.div`
  display: block;
  width: 100%;
  margin-bottom: ${props => (props.last ? "0" : "1")}em;
  position: relative;
`;

const Field = ({ children, last, ...otherProps }) => (
  <StyledField last={last} {...otherProps}>
    {children}
  </StyledField>
);

Field.propTypes = {
  children: PropTypes.node.isRequired,
  last: PropTypes.bool
};

Field.defaultProps = {
  last: false
};

export default Field;
