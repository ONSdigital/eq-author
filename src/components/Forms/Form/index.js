import React from "react";
import PropTypes from "prop-types";
import styled from "styled-components";

const StyledForm = styled.form`
  width: 100%;
  margin-bottom: 0;
`;

const Form = ({ action, children, onSubmit, ...otherProps }) => (
  <StyledForm
    action={action}
    method="POST"
    onSubmit={onSubmit}
    autoComplete="off"
    {...otherProps}
  >
    {children}
  </StyledForm>
);

Form.propTypes = {
  action: PropTypes.string,
  children: PropTypes.node.isRequired,
  onSubmit: PropTypes.func.isRequired
};

export default Form;
