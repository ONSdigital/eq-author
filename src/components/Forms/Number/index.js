import React from "react";
import Input from "components/Forms/Input";
import styled from "styled-components";
import PropTypes from "prop-types";
import { clamp, isNaN } from "lodash";

const StyledDiv = styled.div`
  display: flex;
  flex-direction: row;
`;

const StyledInput = styled(Input)`
  width: 4em;

  &[type="number"]::-webkit-inner-spin-button,
  &[type="number"]::-webkit-outer-spin-button {
    appearance: none;
    margin: 0;
  }

  &[type="number"] {
    appearance: textfield;
  }
`;

class Number extends React.Component {
  state = {
    value: this.props.value
  };

  onComponentDidUpdate() {
    if (this.props.value !== this.state.value) {
      this.setState({
        value: this.props.value
      });
    }
  }

  handleChange = ({ value }) => {
    this.setState({ value });
  };

  handleBlur = () => {
    const name = this.props.name || this.props.id;
    const { value } = this.state;
    const enteredValue = clamp(
      parseInt(value, 10),
      this.props.min,
      this.props.max
    );
    /* eslint-disable no-compare-neg-zero */
    const newValue =
      isNaN(enteredValue) || enteredValue === -0
        ? this.props.min
        : enteredValue;

    this.setState({ value: newValue });

    this.props.onChange({
      name,
      value: newValue
    });
    if (this.props.onBlur) {
      setImmediate(() => {
        this.props.onBlur();
      });
    }
  };

  render() {
    return (
      <StyledDiv>
        <StyledInput
          {...this.props}
          value={this.state.value}
          onChange={this.handleChange}
          type="number"
          onBlur={this.handleBlur}
          aria-live="assertive"
          role="alert"
          data-test="number-input"
        />
      </StyledDiv>
    );
  }
}

Number.defaultProps = {
  min: 0,
  step: 1
};

Number.propTypes = {
  id: PropTypes.string,
  name: PropTypes.string,
  onChange: PropTypes.func.isRequired,
  onBlur: PropTypes.func,
  value: PropTypes.number,
  step: PropTypes.number,
  min: PropTypes.number,
  max: PropTypes.number.isRequired
};

export default Number;
