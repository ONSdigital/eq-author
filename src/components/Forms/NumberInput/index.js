import React from "react";
import Input from "components/Forms/Input";
import styled from "styled-components";

import arrowIcon from "./arrow.svg";

const NumericWrapper = styled.div`
  display: flex;
  flex-direction: row;
`;

const InvertedVertically = styled.img`
  -moz-transform: scaleY(-1);
  -o-transform: scaleY(-1);
  -webkit-transform: scaleY(-1);
  transform: scaleY(-1);
  filter: FlipV;
  -ms-filter: "FlipV";
`;

const StyledNumeric = styled(Input)`
  &[type=number]::-webkit-inner-spin-button, 
  &[type=number]::-webkit-outer-spin-button { 
      -webkit-appearance: none;
      -moz-appearance: none;
      appearance: none;
      margin: 0; 
  }
  
  width: 4em;
`;

const NumericButtons = styled.div`
  display: flex;
  flex-direction: column;
`;

const NumericButton = styled.button`
  border: 0;
  background: none;
`;

class NumberInput extends React.Component {
  constructor() {
    super();
    this.state = {
      value: 0
    };
  }

  componentWillMount = () => {
    if (this.props.value) {
      this.setState({
        value: parseInt(this.props.value, 10).toString()
      });
    }
  };

  handleUp = () => {
    const newValue = parseInt(this.state.value, 10) + 1;
    this.setState({
      value: newValue.toString()
    });
  };

  handleDown = () => {
    const newValue = parseInt(this.state.value, 10) - 1;
    this.setState({
      value: newValue.toString()
    });
  };

  handleChange = ({ value }) => {
    this.setState({
      value: parseInt(value, 10).toString()
    });
  };

  render(props) {
    return (
      <NumericWrapper>
        <StyledNumeric
          type="number"
          value={this.state.value}
          onChange={this.handleChange}
          {...props}
        />
        <NumericButtons>
          <NumericButton type="button" onClick={this.handleUp}>
            <InvertedVertically src={arrowIcon} width="18" />
          </NumericButton>
          <NumericButton type="button" onClick={this.handleDown}>
            <img src={arrowIcon} width="18" />
          </NumericButton>
        </NumericButtons>
      </NumericWrapper>
    );
  }
}

export default NumberInput;
