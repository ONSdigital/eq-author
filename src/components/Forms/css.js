import { css } from "styled-components";
import { colors, radius } from "constants/theme";
import { darken } from "polished";

export const sharedStyles = css`
  padding: 0.7em;
  width: 100%;
  display: block;
  border-radius: ${radius};
  border: 1px solid ${colors.borders};
  font-size: 1em;

  &:hover {
    border-color: ${darken(0.1)(colors.borders)};
  }

  &:focus {
    outline: none;
    border: 1px solid ${colors.primary};
  }
`;
