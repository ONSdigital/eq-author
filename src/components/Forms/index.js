export { default as Form } from "./Form";
export { default as Label } from "./Label";
export { default as Input } from "./Input";
export { default as Select } from "./Select";
export { default as TextArea } from "./TextArea";
export { default as Field } from "./Field";
export { default as Number } from "./Number";
