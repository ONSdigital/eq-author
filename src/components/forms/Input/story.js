import React from 'react';
import { storiesOf, action } from '@kadira/storybook';
import Input from './index'

storiesOf('Input', module)
  .add('without value', () => (
    <Input  />
  ))
  .add('with value', () => (
    <Input value={'hello'} />
  ))
