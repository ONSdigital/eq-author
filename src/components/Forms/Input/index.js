import React from "react";
import PropTypes from "prop-types";
import styled, { css } from "styled-components";
import { sharedStyles } from "components/Forms/css";
import iconCheckbox from "./icon-checkbox.svg";
import withChangeHandler from "components/Forms/withChangeHandler";

const checkBox = css`
  display: inline-block;
  width: 1.1em;
  height: 1.1em;
  padding: 0;
  margin: 0 1em 0 0;
  vertical-align: middle;
  appearance: none;
  font-size: 1em;
  &:checked {
    background: url(${iconCheckbox}) no-repeat center;
    background-size: 0.8em auto;
  }
`;

const StyledInput = styled.input`
  ${sharedStyles};
  ${props => props.type === "checkbox" && checkBox};
`;

const Input = ({ type, defaultValue, id, ...otherProps }) =>
  <StyledInput
    type={type}
    defaultValue={defaultValue}
    id={id}
    name={id}
    {...otherProps}
  />;

Input.defaultProps = {
  type: "text"
};

Input.propTypes = {
  id: PropTypes.string,
  type: PropTypes.oneOf(["text", "checkbox", "radio", "number"]).isRequired,
  defaultValue: PropTypes.oneOfType([
    PropTypes.string,
    PropTypes.bool,
    PropTypes.number
  ])
};

export default withChangeHandler(Input);
