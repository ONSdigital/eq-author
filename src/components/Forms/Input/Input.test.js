import React from "react";
import { mount } from "enzyme";
import { Input } from "components/Forms/Input";

let wrapper;

const defaultValue = "I am some text";

describe("components/Forms/Input", () => {
  beforeEach(() => {
    wrapper = mount(<Input defaultValue={defaultValue} />);
  });

  it("should render correctly", () => {
    expect(wrapper).toMatchSnapshot();
  });

  it("should render a text input by default", () => {
    expect(wrapper.props()).toMatchObject({ type: "text" });
  });

  it("should render the correct type if provided", () => {
    wrapper = mount(<Input type="checkbox" />);
    expect(wrapper.props()).toMatchObject({ type: "checkbox" });
  });

  it("should pass `defaultValue` prop to component when type=text", () => {
    expect(wrapper.props().defaultValue).toEqual(defaultValue);
  });

  it("should pass `defaultChecked` prop to component when type=checkbox", () => {
    wrapper = mount(<Input type="checkbox" defaultChecked />);
    expect(wrapper.find("input").node.checked).toEqual(true);
    wrapper = mount(<Input type="checkbox" defaultChecked={false} />);
    expect(wrapper.find("input").node.checked).toEqual(false);
  });

  it("should call handleChange with appropriate args", () => {
    const handleChange = jest.fn();
    wrapper = mount(<Input id="text" handleChange={handleChange} />);
    wrapper.simulate("change", { target: { value: "hello" } });
    expect(handleChange).toHaveBeenCalledWith({ text: "hello" });
    wrapper = mount(
      <Input
        type="checkbox"
        id="checkbox"
        handleChange={handleChange}
        defaultChecked
      />
    );
    wrapper.simulate("change", { target: { checked: true } });
    expect(handleChange).toHaveBeenCalledWith({ checkbox: true });
  });
});
