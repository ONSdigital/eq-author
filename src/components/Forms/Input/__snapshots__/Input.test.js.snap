// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`components/Forms/Input should render correctly 1`] = `
<Input
  defaultValue="I am some text"
  handleChange={[Function]}
  type="text"
>
  <Input__StyledInput
    defaultValue="I am some text"
    onChange={[Function]}
    type="text"
  >
    <input
      className="Input__StyledInput-gzbIwR cIVRSX"
      defaultValue="I am some text"
      onChange={[Function]}
      type="text"
    />
  </Input__StyledInput>
</Input>
`;
